sonar-core/src/main/java/org/sonar/core/platform/PluginLoader.java

```diff
@@ -45,117 +38,98 @@
       if (Strings.isNullOrEmpty(info.getBasePlugin())) {
-        // The plugins that extend other plugins can only add some files to classloader.
-        // They can't change ordering strategy.
         def.setSelfFirstStrategy(info.isUseChildFirstClassLoader());
-      }
-    }
-    return classloadersByBasePlugin.values();
-  }
-
-  /**
-   * Step 2 - create classloaders with appropriate constituents and metadata
-   */
-  private void buildClassloaders(Collection<ClassloaderDef> defs) {
-    ClassloaderBuilder builder = new ClassloaderBuilder();
-    builder.newClassloader(API_CLASSLOADER_KEY, baseClassloader());
-    for (ClassloaderDef def : defs) {
-      builder
-        .newClassloader(def.getBasePluginKey())
-        .setParent(def.getBasePluginKey(), API_CLASSLOADER_KEY, new Mask())
-        // resources to be exported to other plugin classloaders (siblings)
-        .setExportMask(def.getBasePluginKey(), def.getMask())
-        .setLoadingOrder(def.getBasePluginKey(), def.isSelfFirstStrategy() ? SELF_FIRST : PARENT_FIRST);
-      for (File file : def.getFiles()) {
-        builder.addURL(def.getBasePluginKey(), fileToUrl(file));
-      }
-      for (ClassloaderDef sibling : defs) {
-        if (!sibling.getBasePluginKey().equals(def.getBasePluginKey())) {
-          builder.addSibling(def.getBasePluginKey(), sibling.getBasePluginKey(), new Mask());
+        Version minSqVersion = info.getMinimalSqVersion();
+        boolean compatibilityMode = (minSqVersion != null && minSqVersion.compareToIgnoreQualifier(COMPATIBILITY_MODE_MAX_VERSION) < 0);
+        def.setCompatibilityMode(compatibilityMode);
+        if (compatibilityMode) {
+          Loggers.get(getClass()).info("API compatibility mode is enabled on plugin {} [{}] " +
+            "(built with API lower than {})",
+            info.getName(), info.getKey(), COMPATIBILITY_MODE_MAX_VERSION);
         }
       }
     }
-    Map<String, ClassLoader> classloadersByBasePluginKey = builder.build();
-    for (ClassloaderDef def : defs) {
-      ClassLoader builtClassloader = classloadersByBasePluginKey.get(def.getBasePluginKey());
-      if (builtClassloader == null) {
-        throw new IllegalStateException(String.format("Fail to create classloader for plugin [%s]", def.getBasePluginKey()));
-      }
-      def.setBuiltClassloader(builtClassloader);
-    }
+    return classloadersByBasePlugin.values();
   }
 
   /**
-   * Step 3 - instantiate plugin instances ({@link Plugin}
+   * Instantiates collection of ({@link Plugin} according to given metadata and classloaders
    *
    * @return the instances grouped by plugin key
    * @throws IllegalStateException if at least one plugin can't be correctly loaded
    */
-  private Map<String, Plugin> instantiatePluginInstances(Collection<ClassloaderDef> defs) {
+   @VisibleForTesting
+   Map<String, Plugin> instantiatePluginClasses(Map<PluginClassloaderDef, ClassLoader> classloaders) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Make this line start at column 3. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AIndentationCheck)